pkg/provisioning/service.go

```diff
@@ -234,51 +242,82 @@ func (s *Service) applyConnectorStates(ctx context.Context, pl *pipeline.Instanc
 	return nil
 }
 
-func (s *Service) deletePipeline(ctx context.Context, pl *pipeline.Instance, config PipelineConfig) error {
+func (s *Service) deletePipeline(ctx context.Context, r *rollback.R, pl *pipeline.Instance, config PipelineConfig, connStates map[string]func(context.Context) error) error {
 	var err error
 	// remove pipeline processors
-	for k := range config.Processors {
+	for k, v := range config.Processors {
 		_, err = s.pipelineService.RemoveProcessor(ctx, pl, k)
 		if err != nil {
 			return err
 		}
+		s.rollbackRemoveProcessorFromPipeline(ctx, r, pl, k)
+
 		err = s.processorService.Delete(ctx, k)
 		if err != nil {
 			return err
 		}
+		s.rollbackDeleteProcessor(ctx, r, pl.ID, processor.ParentTypePipeline, k, v)
 	}
+
+	r.Append(func() error {
+		err := s.applyConnectorStates(ctx, pl, connStates)
+		if err != nil {
+			return err
+		}
+		if pl.Status == pipeline.StatusRunning {
+			err = s.pipelineService.Start(ctx, s.connectorService, s.processorService, pl)
+		}
+		return err
+	})
+
 	// remove connector processors
 	for k, cfg := range config.Connectors {
-		for k1 := range cfg.Processors {
+		for k1, cfg1 := range cfg.Processors {
 			_, err = s.connectorService.RemoveProcessor(ctx, k, k1)
 			if err != nil {
 				return err
 			}
+			s.rollbackRemoveProcessorFromConnector(ctx, r, k, k1)
+
 			err = s.processorService.Delete(ctx, k1)
 			if err != nil {
 				return err
 			}
+			s.rollbackDeleteProcessor(ctx, r, k, processor.ParentTypeConnector, k1, cfg1)
 		}
 	}
+
 	// remove connectors
-	for k := range config.Connectors {
+	for k, v := range config.Connectors {
 		_, err = s.pipelineService.RemoveConnector(ctx, pl, k)
 		if err != nil {
 			return err
 		}
+		s.rollbackRemoveConnector(ctx, r, pl, k)
+
 		err = s.connectorService.Delete(ctx, k)
 		if err != nil {
 			return err
 		}
+		s.rollbackDeleteConnector(ctx, r, pl.ID, k, v)
 	}
 	err = s.pipelineService.Delete(ctx, pl)
 	if err != nil {
 		return err
 	}
+	s.rollbackDeletePipeline(ctx, r, pl.ID, config)
```

Comment:
I'm relatively sure that once we reach this line the rest of the rollbacks won't work anymore (or they will produce an unexpected result). See, the way the pipeline service is written is that it expects the caller to supply a `pipeline.Instance` in all calls. That's different from how the connector and processor service work, they expect the caller to supply an ID. So once we delete the pipeline, the rollbacks will work on an old `pipeline.Instance` that doesn't exist anymore.

I don't see an easy fix, we will need to change the interface of the pipeline service to also expect pipeline IDs 😕 the problem is that this PR is already becoming quite huge, so I'm fine with pushing this down the road and adding a comment here describing the problem. We shouldn't forget about this though.

Reply:
was there a reason why the pipeline service was built differently than the others?

Reply:
Honestly, I don't remember. I don't think it was done intentionally.